File: weathersdk/src/main/java/fr/ekito/myweatherlibrary/ws/WeatherMockWS.java
package fr.ekito.myweatherlibrary.ws;

import android.util.Log;

import com.google.gson.Gson;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

import fr.ekito.myweatherlibrary.di.Inject;
import fr.ekito.myweatherlibrary.json.geocode.Geocode;
import fr.ekito.myweatherlibrary.json.geocode.Location;
import fr.ekito.myweatherlibrary.json.weather.Weather;
import rx.Observable;
import rx.Subscriber;

/**
 * Created by arnaud on 14/02/2017.
 */

public class WeatherMockWS implements WeatherWS {

    private static final String TAG = WeatherMockWS.class.getSimpleName();
    private Gson gson = new Gson();
    private Map<Location, String> cities = new HashMap<>();

    public WeatherMockWS() {
        try {
            String[] list = Inject.getApplicationContext().getAssets().list("json");
            for (String fn : list) {
                if (fn.startsWith("geocode_")) {
                    Geocode geocode = gson.fromJson(readJsonFile(fn), Geocode.class);
                    String name = fn.replace("geocode_", "").replace(".json", "");
                    cities.put(geocode.getResults().get(0).getGeometry().getLocation(), name);
                }
            }
            Log.i(TAG, "Can respond for cities : " + cities);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    static String readJsonFile(String jsonFile) {
        StringBuilder buf = new StringBuilder();
        try {
            InputStream json = Inject.getApplicationContext().getAssets().open("json/" + jsonFile);
            BufferedReader in =
                    new BufferedReader(new InputStreamReader(json, "UTF-8"));
            String str;
            while ((str = in.readLine()) != null) {
                buf.append(str);
            }
            in.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return buf.toString();
    }

    @Override
    public Observable<Geocode> geocode(final String address) {
        return Observable.create(new Observable.OnSubscribe<Geocode>() {
            @Override
            public void call(Subscriber<? super Geocode> subscriber) {
                String adrs = address.toLowerCase();
                Geocode geocode;
                if (isKnownCity(adrs)) {
                    geocode = gson.fromJson(readJsonFile("geocode_" + adrs + ".json"), Geocode.class);
                } else {
                    geocode = gson.fromJson(readJsonFile("geocode_toulouse.json"), Geocode.class);
                }
                if (geocode != null) {
                    subscriber.onNext(geocode);
                }
                subscriber.onCompleted();
            }
        });
    }

    private boolean isKnownCity(String adrs) {
        return cities.values().contains(adrs);
    }

    private String cityFromLocation(final Double lat, final Double lng) {
        String city = "toulouse";
        for (Location location : cities.keySet()) {
            if (location.getLat().equals(lat) && location.getLng().equals(lng)) {
                city = cities.get(location);
            }
        }
        return city;
    }

    @Override
    public Observable<Weather> weather(final Double lat, final Double lng, String lang) {
        return Observable.create(new Observable.OnSubscribe<Weather>() {
            @Override
            public void call(Subscriber<? super Weather> subscriber) {
                String city = cityFromLocation(lat, lng);
                Weather weather = gson.fromJson(readJsonFile("weather_" + city + ".json"), Weather.class);
                if (weather != null) {
                    subscriber.onNext(weather);
                }
                subscriber.onCompleted();
            }
        });
    }
}

File: weathersdk/src/main/kotlin/fr/ekito/myweatherlibrary/ws/WeatherMockWS.kt
package fr.ekito.myweatherlibrary.ws

import android.util.Log
import com.google.gson.Gson
import fr.ekito.myweatherlibrary.di.Inject
import fr.ekito.myweatherlibrary.json.geocode.Geocode
import fr.ekito.myweatherlibrary.json.geocode.Location
import fr.ekito.myweatherlibrary.json.weather.Weather
import io.reactivex.Single
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

/**
 * Created by arnaud on 14/02/2017.
 */

val TAG = WeatherMockWS::class.java.simpleName

fun readJsonFile(jsonFile: String): String {
    val buf = StringBuilder()
    val json = Inject.applicationContext()!!.assets.open("json/" + jsonFile)
    // closable resource
    BufferedReader(InputStreamReader(json, "UTF-8")).use { br ->
        buf.append(br.lineSequence().joinToString(separator = "\n"))
    }
    return buf.toString()
}

class WeatherMockWS : WeatherWS {
    private val gson = Gson()
    private val cities = HashMap<Location, String>()

    init {
        val list = Inject.applicationContext()!!.assets.list("json")

        cities += list.filter { it.startsWith("geocode_") }.map {
            val geocode = gson.fromJson<Geocode>(readJsonFile(it), Geocode::class.java)
            val name = it.replace("geocode_", "").replace(".json", "")
            // pair result
            Pair(geocode.results[0].geometry!!.location!!, name)
        }.toMap() // direct to map

        Log.i(TAG, "Can respond for cities : " + cities)
    }

    private fun isKnownCity(adrs: String): Boolean {
        return cities.values.contains(adrs)
    }

    private fun cityFromLocation(lat: Double?, lng: Double?): String {
        var city = "toulouse"
        cities.keys
                .filter { it.lat == lat && it.lng == lng }
                .forEach { city = cities[it]!! }

        return city
    }

    override fun geocode(address: String): Single<Geocode> {
        return Single.create { subscriber ->
            val adrs = address.toLowerCase()
            val geocode: Geocode
            if (isKnownCity(adrs)) {
                geocode = gson.fromJson<Geocode>(readJsonFile("geocode_$adrs.json"), Geocode::class.java)
            } else {
                geocode = gson.fromJson<Geocode>(readJsonFile("geocode_toulouse.json"), Geocode::class.java)
            }
            subscriber.onSuccess(geocode)
        }
    }

    override fun weather(lat: Double?, lon: Double?, lang: String): Single<Weather> {
        return Single.create { subscriber ->
            val city = cityFromLocation(lat, lon)
            val weather: Weather = gson.fromJson<Weather>(readJsonFile("weather_$city.json"), Weather::class.java)
            subscriber.onSuccess(weather)
        }
    }
}
